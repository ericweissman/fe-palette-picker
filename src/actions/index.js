export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
});

export const hasErrored = (message) => ({
  type: 'HAS_ERRORED',
  message
});

//projects
export const getProjectsSuccess = (projects) => ({
  type: 'GET_PROJECTS_SUCCESS',
  projects
})

export const addProjectSuccess = (project) => ({
  type: 'ADD_PROJECT_SUCCESS',
  project
})