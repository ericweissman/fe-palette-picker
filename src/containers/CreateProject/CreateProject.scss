.create-project {
  margin: 10px;
  padding: 10px;
  background-color: $secondary-background;
  h3 {
    color: $text-accent;
    text-align: center;
    margin: 15px;
    font-size: 26px;
  }
  form {
    text-align: center;
  }
  h5 {
    color: $warning-color;
    text-transform: uppercase;
  }
}