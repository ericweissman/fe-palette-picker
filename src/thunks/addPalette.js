import { isLoading, hasErrored, addPaletteSuccess } from '../actions'

export const addPalette = (url, palette) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(palette)
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(addPaletteSuccess(result))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}