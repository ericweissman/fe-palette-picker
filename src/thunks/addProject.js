import { isLoading, hasErrored, addProjectSuccess } from '../actions'

export const addProject = (url, project_name) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type" : "application/json"
        },
        body: JSON.stringify({ project_name })
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(addProjectSuccess(result))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}