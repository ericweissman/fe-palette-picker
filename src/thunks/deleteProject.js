import { isLoading, hasErrored, deleteProjectSuccess } from '../actions'

export const deleteProject = (url, projectID) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: "DELETE",
        headers: {
          "Content-Type" : "application/json"
        }
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      // const result = await response.json()
      dispatch(deleteProjectSuccess(projectID))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}