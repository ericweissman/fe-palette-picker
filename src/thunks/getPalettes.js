import { isLoading, hasErrored, getPalettesSuccess } from '../actions'

export const getPalettes = (url, id) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url)
      if (!response.ok) {
        throw Error(response.statusText)
      }
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(getPalettesSuccess(result, id))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}