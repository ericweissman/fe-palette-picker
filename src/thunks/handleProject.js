import { isLoading, hasErrored } from '../actions'

export const handleProject = (url, actionToDispatch, method, project) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const response = await fetch(url, {
        method: `${method}`,
        headers: {
          "Content-Type": "application/json"
        },
        body: JSON.stringify(project)
      })
      if (!response.ok) {
        throw Error(response.statusText)
      }
      
      dispatch(isLoading(false))
      const result = await response.json()
      dispatch(actionToDispatch(result))
    } catch (error) {
      dispatch(hasErrored(error.message))
    }
  }
}