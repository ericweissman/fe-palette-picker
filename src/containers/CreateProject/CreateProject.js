import React, { Component } from 'react'
import { connect } from 'react-redux'
import { handleProject } from '../../thunks/handleProject'
import { addProjectSuccess } from '../../actions'

export class CreateProject extends Component {
  state = {
    projectName: '',
  }

  handleChange = (e) => {
    const { name, value } = e.target
    this.setState({ [name]: value })
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { handleProject } = this.props
    const project = { project_name: this.state.projectName }
    const url = process.env.REACT_APP_BACKEND_URL + '/api/v1/projects'
    handleProject(url, addProjectSuccess, 'POST', project)
  }

  render() {
    return (
      <div className="create-project">
        <h3>New Project</h3>
        <form>
          <input
            onChange={this.handleChange}
            placeholder='new project name'
            name='projectName'
            value={this.state.projectName}>
          </input>
          <button onClick={this.handleSubmit}>
            Save Project
          </button>
        </form>
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  handleProject: (url, actionToDispatch, method, project) => dispatch(handleProject(url, actionToDispatch, method, project))
})

export default connect(null, mapDispatchToProps)(CreateProject)