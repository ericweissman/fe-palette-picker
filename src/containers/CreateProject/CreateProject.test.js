import React from 'react'
import { shallow } from 'enzyme'
import { handleProject } from '../../thunks/handleProject'
import { addProjectsSuccess } from '../../actions'
import { CreateProject, mapDispatchToProps } from './CreateProject'

jest.mock('../../thunks/handleProject')

describe('CreateProject', () => {
  let wrapper
  const mockProps = {
    handleProject: jest.fn()
  }

  it('should match the correct snapshot', () => {
    wrapper = shallow(<CreateProject {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should have the correct default state', () => {
    const expectedState = { projectName: '', message: '' }
    expect(wrapper.state()).toEqual(expectedState)
  })

  it('should setState when handleChange is called', () => {
    const mockEvent = { target: { name: 'projectName', value: 'New Project' } }
    const expectedState = { projectName: 'New Project', message: '' }
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state()).toEqual(expectedState)
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with handleProject when handleSubmit is called', () => {
      const mockDispatch = jest.fn()
      const mockURL = 'http//localhost:1234'
      const mockProject = { project_name: 'Cool Project ' }
      const mappedProps = mapDispatchToProps(mockDispatch)
      const actionToDispatch = handleProject(mockURL, addProjectsSuccess, 'POST', mockProject)

      mappedProps.handleProject(mockURL, addProjectsSuccess, 'POST', mockProject)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)

    })
  })
})