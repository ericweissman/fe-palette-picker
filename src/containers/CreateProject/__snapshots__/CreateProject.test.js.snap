// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProject should match the correct snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateProject
    handleProject={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Create a New Project
        </h3>,
        <form>
          <input
            autoComplete="off"
            name="projectName"
            onChange={[Function]}
            placeholder="project name"
            value=""
          />
          <button
            onClick={[Function]}
          >
            Save Project
          </button>
          <h5>
            
          </h5>
        </form>,
      ],
      "className": "create-project",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Create a New Project",
        },
        "ref": null,
        "rendered": "Create a New Project",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              autoComplete="off"
              name="projectName"
              onChange={[Function]}
              placeholder="project name"
              value=""
            />,
            <button
              onClick={[Function]}
            >
              Save Project
            </button>,
            <h5>
              
            </h5>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "autoComplete": "off",
              "name": "projectName",
              "onChange": [Function],
              "placeholder": "project name",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Save Project",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Save Project",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "h5",
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Create a New Project
          </h3>,
          <form>
            <input
              autoComplete="off"
              name="projectName"
              onChange={[Function]}
              placeholder="project name"
              value=""
            />
            <button
              onClick={[Function]}
            >
              Save Project
            </button>
            <h5>
              
            </h5>
          </form>,
        ],
        "className": "create-project",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Create a New Project",
          },
          "ref": null,
          "rendered": "Create a New Project",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                autoComplete="off"
                name="projectName"
                onChange={[Function]}
                placeholder="project name"
                value=""
              />,
              <button
                onClick={[Function]}
              >
                Save Project
              </button>,
              <h5>
                
              </h5>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "autoComplete": "off",
                "name": "projectName",
                "onChange": [Function],
                "placeholder": "project name",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Save Project",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Save Project",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
              },
              "ref": null,
              "rendered": "",
              "type": "h5",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
